text/0046-metadata-for-offline-signers.md

```diff
@@ -53,30 +57,38 @@ Metadata should be stripped from parts that are not necessary to parse a signabl
+
 All chain teams are stakeholders, as implementing this feature would require timely effort on their side and would impact compatibility with older tools.
 
 This feature is essential for **all** offline signer tools; many regular signing tools might make use of it. In general, this RFC greatly improves security of any network implementing it, as many governing keys are used with offline signers.
 
 Implementing this RFC would remove requirement to maintain metadata portals manually, as task of metadata verification would be effectively moved to consensus mechanism of the chain.
 
+The idea for this RFC was brought up by runtime implementors and was extensively discussed with offline wallet implementors. It was designed in such a way that it can work easily with the existing offline wallet solutions in the Polkadot ecosystem.
+
 ## Explanation
 
-Detailed description of metadata shortening and digest process is provided in [metadata-shortener](https://github.com/Alzymologist/metadata-shortener) crate (see `cargo doc --open` and examples). Below are presented algorithms of the process.
+he FRAME metadata is providing a lot of information about a FRAME based runtime. It contains information about the pallets, the calls per pallet, the storage entries per pallet, information about runtime apis and type information about most of the types that are used in the runtime. For decoding transactions on an offline wallet we mainly require these type information. Most of the other information in the FRAME metadata are actually not required for the decoding and thus we can remove them. Thus, we have come up with a custom representation of the metadata and how this custom metadata is chunked, ensuring that we only need to send the chunks requiring for decoding a certain transaction to the offline wallet.
 
-### Definitions
+A reference implementation of this process is provided in [metadata-shortener](https://docs.rs/metadata-shortener/latest/metadata_shortener/) crate.
 
+### Definitions
 
 #### Metadata descriptor
 
 Values for:
 
-1. `u8` metadata shortening protocol version, 
-2. `ExtrinsicMetadata`,
-3. `spec_version` `String`,
-4. `spec_name` `String`,
-5. `u16` ss58 prefix,
-6. `u8` decimals value or `0u8` if no units are defined,
-7. `tokenSymbol` `String` defined on chain to identify the name of currency (available for example through `system.properties()` RPC call) or empty string if no base units are defined,
+1. `u8` metadata shortening protocol version (as encoded enum variant), 
+2. Id of registry type encoding all extrinsic calls (u32),
+3. Vector of `SignedExtension` structs as defined in Metadata V15 (vector of named (`String`) pairs of `TypeId` (`u32`))
+4. `spec_version` `String` as found in the `RuntimeVersion` as of generating the metadata. While this information can also be found in the metadata, it is hidden in a big blob of data. To not being required to transfer this big blob of data, we directly add these information here,
```

Comment:
Version is found as `u32` in metadata. I'd stick to the same type. https://github.com/paritytech/polkadot-sdk/blob/53f615de503c8936c2639f51b99019c64eab1018/substrate/primitives/version/src/lib.rs#L185

Reply:
This is dynamic type described in metadata; it is indeed u32 in Polkadot and in Kusama, but it could be anything, probably even enum similar to how Era is constructed - why not? - as long as it is serializable. Something like 

```
struct version {
  major: u32,
  minor: u32,
  notes: String,
}
```

would be reasonable solution.

We've really seen a couple of networks using u16 there. Not that it's all that important, but as long as our protocol would be the only limitation - it would be a limitation for our protocol; let's not enforce it unless it's enforced upstream?

The version is not exactly algebraic number - it has comparison operation, but no other algebra is defined; the primitive variable with these properties is really `String`, not integer. This is silly, we've argued a lot about this internally, but with dumb JS-style solution we seem to have all the flexibility we might want and no real downsides (as in the end from point of view of our protocol, it's just an opaque constant - key for database search).

Again, if this would be a limitation for acceptance of this RFC, it's not worth it, let's set it to whatever.

Reply:
Let's go for `String` then, I like the rationale behind it. Could you write it there?

Reply:
done, check it out